src/Language/Futhark/Interpreter/AD.hs

```diff
@@ -0,0 +1,289 @@
+    Int64 -> IntValue $ Int64Value $ round c
+  FloatType t' -> case t' of
+    Float16 -> FloatValue $ Float16Value $ fromRational c
+    Float32 -> FloatValue $ Float32Value $ fromRational c
+    Float64 -> FloatValue $ Float64Value $ fromRational c
+  _ -> error $ "No valid prim value for " ++ show t
+
+doOp :: Op -> [ADValue] -> Maybe ADValue
+doOp op p = do
+  let p' = map value p
+  if opTypeMatch op p' then do
+    v <- case (op, p') of
+      (OpBin  op', [x, y]) -> doBinOp op' x y
+      (OpCmp  op', [x, y]) -> BoolValue <$> doCmpOp op' x y
+      (OpUn   op', [x])    -> doUnOp op' x
+      (OpConv op', [x])    -> doConvOp op' x
+      (OpFn   fn,  _)      -> do
+        (_, _, f) <- M.lookup fn primFuns
+        f p'
+      _ -> error "Not implemented"
+
+    case op of
+      (OpBin _)  -> handleOp op p v
+      (OpCmp _)  -> Just $ Primal v
+      (OpUn _)   -> handleOp op p v
+      (OpFn _)   -> handleOp op p v
+      (OpConv _) -> Nothing
+
+  else Nothing
+
+deriveOp :: Op -> [PrimExp VName] -> Maybe [PrimExp VName]
+deriveOp (OpBin op) [x, y] = Just $ do
+  let (a, b) = pdBinOp op x y
+  [a, b]
+deriveOp (OpUn  op) [x] = Just [pdUnOp op x]
+deriveOp (OpFn  fn) p   = pdBuiltin (nameFromString fn) p
+deriveOp _ _ = Nothing
+
+handleOp :: Op -> [ADValue] -> PrimValue -> Maybe ADValue
+handleOp op p v = do
+  -- Find the maximum depth
+  let pDepth = map (\case
+        Primal _ -> 0
+        Seed s -> depth s) p
+  let maxDepth = maximum pDepth
+
+  if maxDepth == 0 then Just $ Primal v
+  else do
+    -- Get all the top level parameters
+    let d = map (\v' -> case v' of
+          Primal _ -> Left v'
+          Seed se -> if depth se == maxDepth then Right se
+                                             else Left v') p
+
+    -- Handle the previous level
+    let d' = map (\case
+          Right se -> primal se
+          Left v' -> v') d
+    v' <- handleOp op d' v
+
+    -- Call the top level handler
+    case find (\case Right _ -> True
+                     Left _  -> False) d of
+      Just (Right (VjpSeed {})) -> Seed . VjpSeed maxDepth <$> vjpHandleFn op d v'
+      Just (Right (JvpSeed {})) -> jvpHandleFn op d maxDepth v'
+      _ -> Just $ Primal v -- ?TODO: This never happens, so maybe remove it in some way?
+
+runPrimExp :: PrimExp VName -> M.Map VName ADValue -> Maybe ADValue
+runPrimExp (LeafExp n _) m = M.lookup n m
+runPrimExp (ValueExp pv) _ = Just $ Primal pv
+runPrimExp (BinOpExp op x y) m = do
+  x' <- runPrimExp x m
+  y' <- runPrimExp y m
+  doOp (OpBin op) [x', y']
+runPrimExp (CmpOpExp op x y) m = do
+  x' <- runPrimExp x m
+  y' <- runPrimExp y m
+  doOp (OpCmp op) [x', y']
+runPrimExp (UnOpExp op x) m = do
+  x' <- runPrimExp x m
+  doOp (OpUn op) [x']
+runPrimExp (ConvOpExp op x) m = do
+  x' <- runPrimExp x m
+  doOp (OpConv op) [x']
+runPrimExp (FunExp fn p _) m = do
+  p' <- mapM (`runPrimExp` m) p
+  doOp (OpFn fn) p'
+
+
+-- VJP
+
+data Tape
+  = TapeId Int ADValue
```

Comment:
You convinced me that the `Int` here is necessary (and it is obvious in retrospect), but it is *really* subtle. I suggest coming up with a name for it (too bad "seed" is taken - or is it?).